homeassistant/components/cloud/http_api.py

```diff
@@ -19,59 +21,66 @@ def async_setup(hass):
+    auth_api.ExpiredCode: (400, 'Confirmation code has expired.'),
+    auth_api.InvalidCode: (400, 'Invalid confirmation code.'),
+    asyncio.TimeoutError: (502, 'Unable to reach the Home Assistant cloud.')
+}
 
-    url = '/api/cloud/login'
-    name = 'api:cloud:login'
-    schema = vol.Schema({
-        vol.Required('username'): str,
-        vol.Required('password'): str,
-    })
 
+def _handle_cloud_errors(handler):
+    """Helper method to handle auth errors."""
     @asyncio.coroutine
-    def post(self, request):
-        """Validate config and return results."""
+    @wraps(handler)
+    def error_handler(view, request, *args, **kwargs):
+        """Handle exceptions that raise from the wrapped request handler."""
         try:
-            data = yield from request.json()
-        except ValueError:
-            _LOGGER.error('Login with invalid JSON')
-            return self.json_message('Invalid JSON.', 400)
+            result = yield from handler(view, request, *args, **kwargs)
+            return result
 
-        try:
-            self.schema(data)
-        except vol.Invalid as err:
-            _LOGGER.error('Login with invalid formatted data')
-            return self.json_message(
-                'Message format incorrect: {}'.format(err), 400)
+        except (auth_api.CloudError, asyncio.TimeoutError) as err:
+            err_info = _CLOUD_ERRORS.get(err.__class__)
+            if err_info is None:
+                err_info = (502, 'Unexpected error: {}'.format(err))
+            status, msg = err_info
+            return view.json_message(msg, status_code=status,
+                                     message_code=err.__class__.__name__)
 
-        hass = request.app['hass']
-        phase = 1
-        try:
-            with async_timeout.timeout(REQUEST_TIMEOUT, loop=hass.loop):
-                cloud = yield from cloud_api.async_login(
-                    hass, data['username'], data['password'])
+    return error_handler
 
-            phase += 1
 
-            with async_timeout.timeout(REQUEST_TIMEOUT, loop=hass.loop):
-                yield from cloud.async_refresh_account_info()
+class CloudLoginView(HomeAssistantView):
+    """Login to Home Assistant cloud."""
 
-        except cloud_api.Unauthenticated:
-            return self.json_message(
-                'Authentication failed (phase {}).'.format(phase), 401)
-        except cloud_api.UnknownError:
-            return self.json_message(
-                'Unknown error occurred (phase {}).'.format(phase), 500)
-        except asyncio.TimeoutError:
-            return self.json_message(
-                'Unable to reach Home Assistant cloud '
-                '(phase {}).'.format(phase), 502)
+    url = '/api/cloud/login'
+    name = 'api:cloud:login'
 
-        hass.data[DOMAIN]['cloud'] = cloud
-        return self.json(cloud.account)
+    @asyncio.coroutine
+    @_handle_cloud_errors
+    @RequestDataValidator(vol.Schema({
```

Comment:
I think we should defined the Schema not inline. Maybe in `cloud/const.py` or `cloud/schema.py`

Reply:
but it's only used right here? Defining it in const makes it seem like we plan on re-using it, which is not the case.

Reply:
I kinda like that it is defined close to where it is used, there is no need to look around for it.